src/udev/udev-builtin-usb_id.c

```diff
@@ -177,9 +179,12 @@ static int dev_if_packed_info(sd_device *dev, char *ifs_str, size_t len) {
                 struct usb_interface_descriptor *desc;
                 char if_str[8];
 
-                desc = (struct usb_interface_descriptor *) &buf[pos];
+                desc = (struct usb_interface_descriptor *) (buf + pos);
                 if (desc->bLength < 3)
                         break;
+                if (desc->bLength > size - sizeof(struct usb_interface_descriptor))
+                        return log_device_debug_errno(dev, SYNTHETIC_ERRNO(EIO),
+                                                      "Corrupt data read from \"%s\"", filename);
```

Comment:
the while() loop already does a similar check really, hence the vulnerability doesn't really exist...

I wonder if it's a good idea to validate bLength too closely here. It appears it would be something that is extensible in theory at least and I am pretty sure lots of usb hw won't initialize this correctly. Hence I think it would be wise here to do what the old code did: use what we can use, and don't complain too loud or exit early when the data is otherwise bogus.

Reply:
We use bLength to calculate the next position. This calculation could overflow, and the `while` check is later.

bLength tells us where to read next. If this value is bogus, we'll give bogus output. So I think we have to verify it.

Reply:
i have the suspicion this will bite us in the ass sooner or later. But let's merge.

Reply:
i think the ternary operator would be prettier and safer...

Reply:
(the overflow thing matters indeed)